File: mac/Sources/HexBridgePairing/PairingSeal.swift
import CommonCrypto
import CryptoKit
import Foundation

/// Unwraps the pairing URI the PC hands over in exchange for the short code.
///
/// The exchange used to answer in plain text over plain HTTP. Across a home network that
/// was a considered trade; across the internet it hands the 32-byte key to anyone on the
/// path, and macOS refused the request outright rather than let that happen. The two
/// screens already share a secret — the twelve characters being copied across — so that is
/// what the answer is encrypted with, and no exception to anybody's transport policy is
/// needed.
///
/// **Why the derivation is slow.** The code is twelve characters from a 32-symbol
/// alphabet: exactly 60 bits. Against somebody who captured the blob and can then guess
/// offline, 60 bits behind a fast KDF is worth about a weekend of rented GPUs.
/// PBKDF2-SHA256 at ``iterations`` multiplies that by roughly 2²⁰, which puts it well past
/// the value of one microphone link. It costs each side a single derivation during a step
/// the person is already waiting on.
///
/// This lives in its own target rather than in the app for the same reason
/// `HexBridgeDiscovery` does: it is a security decision, so it has to be testable, and the
/// app target cannot be linked into a test bundle.
///
/// The other half is `PairingSeal` in `win/src/HexBridge.Core/PairingSeal.cs`. Both are
/// pinned to the same vector in `docs/PROTOCOL.md`.
public enum PairingSeal {
    /// Leading byte of the blob, and the only associated data the tag covers.
    public static let version: UInt8 = 1

    public static let saltBytes = 16
    public static let nonceBytes = 12
    public static let tagBytes = 16

    /// PBKDF2-SHA256 rounds. See the note above about why this is not HKDF.
    public static let iterations = 600_000

    /// The 32 symbols a code is written in. I, O, 0 and 1 are absent: this is a string
    /// somebody reads off one screen and types into another.
    public static let alphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

    /// Upper case, alphabet only. Matches `ShortCode.Normalise` on the PC exactly — the
    /// key is derived from the result, so a difference of one character is a silent
    /// failure to decrypt rather than a mismatch anybody could read.
    public static func normalise(_ code: String) -> String {
        String(code.uppercased().filter { alphabet.contains($0) })
    }

    /// The key both sides arrive at from the same code and salt.
    public static func deriveKey(code: String, salt: Data) -> SymmetricKey {
        let password = Array(normalise(code).utf8)
        var derived = [UInt8](repeating: 0, count: 32)

        let status = salt.withUnsafeBytes { saltBytes -> Int32 in
            CCKeyDerivationPBKDF(
                CCPBKDFAlgorithm(kCCPBKDF2),
                // The password is ASCII by construction: it came through `normalise`.
                password.withUnsafeBytes { $0.baseAddress!.assumingMemoryBound(to: CChar.self) },
                password.count,
                saltBytes.baseAddress!.assumingMemoryBound(to: UInt8.self),
                salt.count,
                CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                UInt32(iterations),
                &derived,
                derived.count
            )
        }

        // The only documented failure is a bad parameter, and every parameter here is a
        // constant or a length. A zero key would be worse than a crash.
        precondition(status == kCCSuccess, "PBKDF2 refused fixed parameters")
        return SymmetricKey(data: Data(derived))
    }

    /// Seals a payload. The PC does this in production; the Mac needs it to check itself
    /// against the shared vector.
    public static func seal(_ plaintext: String, code: String, salt: Data, nonce: Data) -> String? {
        seal(plaintext, key: deriveKey(code: code, salt: salt), salt: salt, nonce: nonce)
    }

    public static func seal(_ plaintext: String, key: SymmetricKey, salt: Data, nonce: Data) -> String? {
        guard salt.count == saltBytes, nonce.count == nonceBytes,
              let boxNonce = try? AES.GCM.Nonce(data: nonce),
              let box = try? AES.GCM.seal(
                  Data(plaintext.utf8), using: key, nonce: boxNonce, authenticating: Data([version])
              )
        else { return nil }

        var blob = Data([version])
        blob.append(salt)
        blob.append(nonce)
        blob.append(box.ciphertext)
        blob.append(box.tag)
        return blob.base64EncodedString()
    }

    /// The payload inside the blob, or nil if it is malformed, of a version this build
    /// does not know, or sealed under a different code.
    public static func open(_ blob: String, code: String) -> String? {
        guard let raw = Data(base64Encoded: blob.trimmingCharacters(in: .whitespacesAndNewlines)),
              raw.count >= 1 + saltBytes + nonceBytes + tagBytes,
              raw[raw.startIndex] == version
        else { return nil }

        // Data slices keep the parent's indices, so every bound is written from the start
        // rather than from zero.
        let base = raw.startIndex
        let saltEnd = base + 1 + saltBytes
        let nonceEnd = saltEnd + nonceBytes
        let cipherEnd = raw.endIndex - tagBytes

        let salt = Data(raw[(base + 1)..<saltEnd])
        let nonce = Data(raw[saltEnd..<nonceEnd])
        let cipher = Data(raw[nonceEnd..<cipherEnd])
        let tag = Data(raw[cipherEnd..<raw.endIndex])

        guard let boxNonce = try? AES.GCM.Nonce(data: nonce),
              let box = try? AES.GCM.SealedBox(nonce: boxNonce, ciphertext: cipher, tag: tag),
              let plain = try? AES.GCM.open(
                  box, using: deriveKey(code: code, salt: salt), authenticating: Data([version])
              )
        else { return nil }

        return String(data: plain, encoding: .utf8)
    }
}

File: win/src/HexBridge.Core/PairingSeal.cs
using System.Security.Cryptography;
using System.Text;

namespace HexBridge;

/// <summary>
/// Wraps the pairing URI so the short-code exchange can cross a network nobody trusts.
///
/// <para>
/// The exchange used to answer with the URI as plain text over plain HTTP. On a home
/// network that was a considered trade; across the internet it hands the 32-byte key to
/// anyone on the path, and macOS refused the request outright rather than let it happen.
/// The two screens already share a secret — the twelve characters the person is copying
/// across — so that is what the answer is encrypted with, and no exception to anybody's
/// transport policy is needed.
/// </para>
///
/// <para>
/// <b>Why a slow derivation.</b> The code is twelve characters from a 32-symbol alphabet:
/// exactly 60 bits. Against someone who captured the blob and can guess offline, 60 bits
/// of HKDF would be worth roughly a weekend of rented GPUs. PBKDF2-SHA256 at
/// <see cref="Iterations"/> multiplies that by about 2^19, which puts it out of reach for
/// the value of one microphone link — and costs each side one derivation, once, during a
/// step a person is already waiting on. The PC derives when it *shows* the code rather
/// than per request, so an attacker cannot spend its CPU by asking repeatedly; and the
/// code has to match before there is anything to decrypt at all.
/// </para>
///
/// <para>
/// The Mac's half is <c>PairingSeal</c> in <c>mac/Sources/HexBridge/Core/Pairing.swift</c>.
/// Both are pinned to the same vector in <c>docs/PROTOCOL.md</c> — change one and the
/// tests on the other platform fail, which is the point.
/// </para>
/// </summary>
public static class PairingSeal
{
    /// <summary>Leading byte of the blob, and the only associated data the tag covers.</summary>
    public const byte Version = 1;

    public const int SaltBytes = 16;
    public const int NonceBytes = 12;
    public const int TagBytes = 16;

    /// <summary>PBKDF2-SHA256 rounds. See the note on the class about why this is not HKDF.</summary>
    public const int Iterations = 600_000;

    /// <summary>
    /// The only associated data the tag covers: the version byte. Held as a field because
    /// a collection expression at the call site leaves the AesGcm overload ambiguous.
    /// </summary>
    private static readonly byte[] AssociatedData = [Version];

    /// <summary>
    /// The key both sides arrive at from the same code and salt.
    ///
    /// <paramref name="code"/> is normalised first, so it does not matter whether it came
    /// from a screen with hyphens or from a text field mid-edit.
    /// </summary>
    public static byte[] DeriveKey(string code, byte[] salt) =>
        Rfc2898DeriveBytes.Pbkdf2(
            Encoding.ASCII.GetBytes(ShortCode.Normalise(code)),
            salt,
            Iterations,
            HashAlgorithmName.SHA256,
            32);

    private static readonly byte[] RendezvousPrefix = "hexbridge-pair-rendezvous-v1"u8.ToArray();

    /// <summary>
    /// The name two machines that know the same code agree on without either of them
    /// saying it, so a relay can introduce them without being told the code.
    ///
    /// <para>
    /// It has to be a one-way function of the code and nothing else. The relay holds the
    /// sealed answer, and the key to that answer is derived from the code — so a relay
    /// that learned the code could read what it is carrying, which is the one thing this
    /// relay is built not to do. Guessing the id means guessing the code: 60 bits.
    /// </para>
    ///
    /// <para>
    /// Matched by <c>RendezvousID</c> in <c>relay/rendezvous.go</c> and
    /// <c>PairingSeal.rendezvousID</c> on the Mac, against one vector in docs/PROTOCOL.md.
    /// </para>
    /// </summary>
    public static string RendezvousId(string code)
    {
        var material = Encoding.ASCII.GetBytes(ShortCode.Normalise(code));
        var input = new byte[RendezvousPrefix.Length + material.Length];
        RendezvousPrefix.CopyTo(input, 0);
        material.CopyTo(input, RendezvousPrefix.Length);

        // Base64url without padding, truncated to 16 bytes: 22 characters, and the same
        // shape as the discovery tag so one reader recognises both.
        return Convert.ToBase64String(SHA256.HashData(input).AsSpan(0, 16))
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=');
    }

    /// <summary>Seals <paramref name="plaintext"/> under a fresh salt and nonce.</summary>
    public static string Seal(string plaintext, string code)
    {
        var salt = RandomNumberGenerator.GetBytes(SaltBytes);
        var nonce = RandomNumberGenerator.GetBytes(NonceBytes);
        return Seal(plaintext, DeriveKey(code, salt), salt, nonce);
    }

    /// <summary>
    /// Seals with a caller's salt and nonce, and with the key already derived.
    ///
    /// Used by the wizard, which derives once when it starts showing a code, and by the
    /// tests, which need the answer to be reproducible.
    /// </summary>
    public static string Seal(string plaintext, byte[] key, byte[] salt, byte[] nonce)
    {
        ArgumentOutOfRangeException.ThrowIfNotEqual(salt.Length, SaltBytes);
        ArgumentOutOfRangeException.ThrowIfNotEqual(nonce.Length, NonceBytes);

        var body = Encoding.UTF8.GetBytes(plaintext);
        var cipher = new byte[body.Length];
        var tag = new byte[TagBytes];

        using var gcm = new AesGcm(key, TagBytes);
        gcm.Encrypt(nonce, body, cipher, tag, AssociatedData);

        var blob = new byte[1 + SaltBytes + NonceBytes + cipher.Length + TagBytes];
        blob[0] = Version;
        salt.CopyTo(blob, 1);
        nonce.CopyTo(blob, 1 + SaltBytes);
        cipher.CopyTo(blob, 1 + SaltBytes + NonceBytes);
        tag.CopyTo(blob, 1 + SaltBytes + NonceBytes + cipher.Length);

        return Convert.ToBase64String(blob);
    }

    /// <summary>
    /// Undoes <see cref="Seal(string, string)"/>, or null if the blob is malformed, the
    /// version is one this build does not know, or the code is wrong.
    ///
    /// Windows never needs this in production — it is the side that seals — but the two
    /// halves of a format are only really pinned when one implementation can check itself.
    /// </summary>
    public static string? Open(string blob, string code)
    {
        byte[] raw;
        try
        {
            raw = Convert.FromBase64String(blob.Trim());
        }
        catch (FormatException)
        {
            return null;
        }

        if (raw.Length < 1 + SaltBytes + NonceBytes + TagBytes) return null;
        if (raw[0] != Version) return null;

        var salt = raw[1..(1 + SaltBytes)];
        var nonce = raw[(1 + SaltBytes)..(1 + SaltBytes + NonceBytes)];
        var cipher = raw[(1 + SaltBytes + NonceBytes)..^TagBytes];
        var tag = raw[^TagBytes..];

        var plain = new byte[cipher.Length];
        try
        {
            using var gcm = new AesGcm(DeriveKey(code, salt), TagBytes);
            gcm.Decrypt(nonce, cipher, tag, plain, AssociatedData);
        }
        catch (CryptographicException)
        {
            return null;
        }

        return Encoding.UTF8.GetString(plain);
    }
}
